### ISSUE TITLE
Quick question regarding Equation 2 from paper

### ISSUE THREAD

**Issue by Divadi** [crockwell/rel_pose#4]
Thank you for the interesting work. I just had a quick question regarding Equation 2 of the paper:
![image](https://user-images.githubusercontent.com/17109672/202954513-acfbea22-3fc9-4a52-ae36-09a67fbf1762.png)
Previously, it is explained that $Q_1$ and $K_2$ are both $P \times D$ matrices, where $P$ is the # of patches and $D$ is the feature dimension. Then, $\text{norm}(Q_1K_2^T)$ should be of dimension $P \times P$, with the first $P$ indicating image 1 patches and the latter indicating image 2 patches. However, it seems odd to me that the $V_2$ is also matrix-multiplied on the left side, effectively matching the image 2 patches' features of $V_2$ with the image 1's attention rows of $\text{norm}(Q_1K_2^T)$. Further, if the two images had different # of patches, this operation would not work either. Could I get some insight as to why this is done? I apologize if I have misinterpreted something.

**Comment by crockwell** [comment]
Hey Jinhyung,

Thanks for the question, and apologies for the delay. 

- Number of patches not equal: You make a good point -- indeed if the number of patches in each image weren't equal, this operation wouldn't work. I would argue  assuming the number of patches is equal is actually pretty reasonable in practice, given the patches are specified by the user, and pose is typically predicted between images of similar shape. But agreed, it is indeed a limitation.

- At first glance, I agree it feels unintuitive to multiply V2 on both sides of A. For this I direct you to https://github.com/crockwell/rel_pose/issues/2, where I go into some detail. But please feel free to let me know if you still have questions after reading this!

**Comment by triangleCZH** [comment]
Hi, I have a similar question.
Firstly, I want to make sure I understand this paper correctly:
You use the three tricks to make the attention result explicitly contain `Phi A Phi^T`, which is equivalent to `U^T U` (which, in 8-point algorithm, the essential matrix can be derived from). 

Secondly, I saw you use the same `Phi` for `Phi A Phi ^ T`, and in the code that you also use the same positional encoding for the two images
`
v1 = torch.cat([v1,positional.unsqueeze(1).repeat(1,self.num_heads,1,1)],dim=3)

v2 = torch.cat([v2,positional.unsqueeze(1).repeat(1,self.num_heads,1,1)],dim=3)
`

I think this only hold when 1) v1 and v2 have same patch_h and patch_w, 2) they have the same intrinsics, is this correct?
So I feel like `Phi A Phi ^ T` might be abused by a little bit, in general it should be `Phi A Phi' ^ T` to indicate different images.
This might be trivial, it just confused me a bit.

**Comment by crockwell** [comment]
Yep, you're totally right -- good point! We assume we have the same patch setup on both images and same intrinsics, although it may have been somewhat implicit.
